### ISSUE TITLE
Uglifyjs2 update (2.18) breaks uglifyfication with mangle:true with angular.js code!

### ISSUE THREAD

**Issue by alonisser** [gruntjs/grunt-contrib-uglify#313]
Can't figure out exactly why but installing 2.17 instead works fine

Please FIX the uglifyjs2 version without tilda and fixed version in your package json,  PR coming. since all the js world is currently broken :)

**Comment by arthurvr** [comment]
Is there a ticket about this in the actual uglify repo?

**Comment by jrhite** [comment]
I don't know what the issue is, but there was a new 2.4.18 and now 2.4.19
released yesterday. If you were already using uglifyjs to obfuscate angular
code and it was working fine prior to yesterday, there might be a real bug.
Please file that here:

https://github.com/mishoo/UglifyJS2

Just to be clear though, in case you aren't already aware, if you're using
grunt-contrib-uglify with angular code you should probably be using
ng-annotate too, because obfuscation can break on angular's dependency
injection system. If you don't use ng-annotate see
https://docs.angularjs.org/tutorial/step_05 and search for 'A Note on
Minification' on how to write angular code in an obfuscateable way.




On Mon, Mar 30, 2015 at 1:25 AM, Arthur Verschaeve <notifications@github.com

**Comment by alonisser** [comment]
Yes, the breaking change is 2.4.18 , and yes It's concerning a specific ng
annotation case that worked with the previous version but not with the new
versions, I did fix our code not to fail, but since ugilfyjs2 is presenting
breaking changes in a patch version I still think a widely used plugin as
grunt-contrib-uglify should use fixed versions and bump uglifyjs2 versions
only after checking



Twitter:@alonisser <https://twitter.com/alonisser>
LinkedIn Profile <http://www.linkedin.com/in/alonisser>
Facebook <https://www.facebook.com/alonisser>
*Tech blog:*4p-tech.co.il/blog
*Personal Blog:*degeladom.wordpress.com
Tel:972-54-6734469

**Comment by mishoo** [comment]
@alonisser I can only advise you to [file an issue](https://github.com/mishoo/UglifyJS2/issues) at UglifyJS, if you feel the breaking change comes from there.  Please include a small test case which does *not* involve Grunt.

The idea is that @jrhite sent a PR for grunt-contrib-uglify which actually depends on a feature introduced in UglifyJS 2.4.18, so sooner or later it will have to upgrade.

**Comment by alonisser** [comment]
I see,



Twitter:@alonisser <https://twitter.com/alonisser>
LinkedIn Profile <http://www.linkedin.com/in/alonisser>
Facebook <https://www.facebook.com/alonisser>
*Tech blog:*4p-tech.co.il/blog
*Personal Blog:*degeladom.wordpress.com
Tel:972-54-6734469

**Comment by alonisser** [comment]
I think a better idea would be to accept the pr for this version and merge
@jrhite PR in the next version of grunt-contrib-uglify, so users who suffer
from this problem won't have their build process broken



Twitter:@alonisser <https://twitter.com/alonisser>
LinkedIn Profile <http://www.linkedin.com/in/alonisser>
Facebook <https://www.facebook.com/alonisser>
*Tech blog:*4p-tech.co.il/blog
*Personal Blog:*degeladom.wordpress.com
Tel:972-54-6734469

**Comment by jrhite** [comment]
Yes, totally agreed @alonisser. I'll +1 your PR. It will be nice if someone
picks it up and commits it quickly.

As you say, that will lock in people with the previous code and they'll be
safe. Then moving forward we can work out any kinks in the new updates.

Cheers.

**Comment by mishoo** [comment]
@alonisser I do agree pinning down packages to a known working version is generally a good idea. However, if you see me sticking around and willing to help, it would be an even better idea to point out what is the actual problem with UglifyJS, rather than pushing for a quick fix that will break anyway in the long run.

**Comment by alonisser** [comment]
@mishoo First of all I want to thank you for uglifyjs2! It's a great project and has an important part in our js world.  now for the bug:

I don't exactly know what the bug is, it has something to do with mangle behavior with angular.js when not using the (recommended) [] notation  for an "unused" injected service (that's why it didn't break until now) . when I override grunt-contrib-uglify version of uglifyJS to 2.4.17 everything passes fine with my build process, but when bumping to 18 (or 19) the build breaks. we did manage to fix our code, but we spent half a day going through dependencies and the build process to find this.. and I guess lots of people are going through the same pain.
